src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java

```diff
@@ -32,96 +41,123 @@
    */
   private final GitHubDataCache<GHRepository> repositoryCache = new GitHubDataCache<>();
 
   /**
-   * A private constructor for singleton pattern.
+   * A cache of local repositories.
    */
-  private GitHubDataFetcher() {
-    // private constructor
-  }
+  final LRUMap<GitHubProject, LocalRepository> localRepositories
+      = new LRUMap<>(LOCAL_REPOSITORIES_CACHE_CAPACITY, SCAN_UNTIL_REMOVABLE);
 
   /**
-   * Creates a single instance of this helper class.
-   * 
-   * @return instance of this data pull helper.
+   * A base directory.
    */
-  public static GitHubDataFetcher instance() {
-    return INSTANCE;
-  }
-  
+  private final Path base;
+
   /**
-   * Gets the list of commits for a GitHub project repository using GitHub API. Then, the commit
-   * list is sorted in reverse order by commit date. This will ensure that the commit list will hold
-   * data from latest commit to the first commit. This sorted list will then be stored in a cache
-   * ({@link LRUMap}).
-   * 
-   * @param project of type {@link GitHubProject}, which holds the project information.
-   * @param github of type {@link GitHub} connection to GitHub API.
-   * @return List of {@link GHCommit}.
-   * @throws IOException occurred during REST call to GitHub API.
+   * Info about local repositories.
    */
-  public List<GHCommit> commitsFor(GitHubProject project, GitHub github) throws IOException {
-    Optional<List<GHCommit>> cachedCommits = commitsCache.get(project);
-    if (cachedCommits.isPresent()) {
-      return cachedCommits.get();
-    }
+  private final Map<URL, LocalRepositoryInfo> localRepositoriesInfo;
 
-    List<GHCommit> commits = new ArrayList<>(commitsFor(repositoryFor(project, github)));
-    commits.sort(BY_COMMIT_DATE.reversed());
-    commitsCache.put(project, commits, expiration());
-    return commits;
+  /**
+   * Defines how often new updates should be pulled to a local repository.
+   */
+  private Duration pullInterval = DEFAULT_PULL_INTERVAL;
+
+  /**
+   * Initializes a new data fetcher.
+   *
+   * @param github An interface to the GitHub API.
+   * @throws IOException If something went wrong.
+   */
+  public GitHubDataFetcher(GitHub github) throws IOException {
+    this(github, Paths.get(DEFAULT_DIRECTORY));
   }
 
   /**
-   * Gets the list of commits for a GitHub project repository using GitHub API.
-   * 
-   * @param repository The {@link GHRepository}.
-   * @return List of {@link GHCommit}.
-   * @throws IOException occurred during REST call to GitHub API.
+   * Returns the interface to the GitHub API.
    */
-  private List<GHCommit> commitsFor(GHRepository repository) throws IOException {
-    try {
-      return repository.listCommits().toList();
-    } catch (HttpException e) {
-      LOGGER.error(String.format("Could not fetch commits from %s", repository.getUrl()), e);
-      return Collections.emptyList();
+  public GitHub github() {
+    return github;
```

Comment:
Could this ever return a null?

Reply:
Nope. The field is final and there is a null-check in the constructor.

Reply:
ok